fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java

```diff
@@ -708,8 +708,16 @@ private KeyExpression indexKeyExpressionForPlan(@Nullable KeyExpression commonPr
             final List<KeyExpression> keys = new ArrayList<>(commonPrimaryKey.normalizeKeyForPositions());
             index.trimPrimaryKey(keys);
             if (!keys.isEmpty()) {
-                keys.add(0, indexKeyExpression);
-                indexKeyExpression = Key.Expressions.concat(keys);
+                final List<KeyExpression> children = new ArrayList<>(keys.size() + 1);
+                children.add(indexKeyExpression);
+                for (KeyExpression key : keys) {
+                    if (key.getColumnSize() == 1) {
+                        children.add(key);
+                    } else {
+                        children.add(Key.Expressions.list(key));
+                    }
+                }
+                indexKeyExpression = Key.Expressions.concat(children);
```

Comment:
I see why this is safer in a particularly sensitive area. So maybe this fix needs to leave a comment here about the specially-normalized form it is producing. Particularly like 717, which is kind of putting back a list (or something like that).

Reply:
I went to add the comment, and I did a bit more thinking. There was another edge case that I worried me, namely that if the primary key was only a single column, with the current normalization, we could end up issuing an incorrect comparison on that field because this normalization doesn't remember if single column expressions are serialized into nested tuples or not. So, I think we need to keep track of that, and the more I thought about it, the more it made sense to change `ListKeyExpression`.

I've put that in its own commit (f86e1dc). (The rebase allows me to pick up some randomized test utilities from `main`.) I was initially worried that this would affect union comparison keys more broadly, but I now realize that it will only affect synthetic record types (or indexes on `ListKeyExpression`s, I guess, but that seems less likely). So, given the new-ness of the type, I _think_ the right thing is to call this out in the release notes, though maybe there's a way we could get the old behavior back as a temporary-ish thing.

Reply:
I think accepting that some plans / plan-hashes will change for unions of synthetic types is probably acceptable. Both because they are new and because such plans are the ones being fixed by this change.

I thought that perhaps some covering cases where a required field was a constituent primary key would also be affected. We don't seem to have any tests like that. Is that a valid concern?